### PR TITLE
test/journal: always close object

### PR DESCRIPTION
**Pull request by tchaikov**
otherwise when we destruct `TestObjectRecorder::Handler::lock` or
`journal::JournalRecorder::m_object_locks`, it/they would be still
being waited by some condition variable.

Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
this addresses the failure like
```
[ RUN      ] TestObjectRecorder.AppendFilledObject
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/common/mutex_debug.h: In function 'ceph::mutex_debug_detail::mutex_debug_impl<<anonymous> >::~mutex_debug_impl() [with bool Recursive = false]' thread ffff82770000 time 2019-08-04T02:22:12.630469-0400
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/common/mutex_debug.h: 111: FAILED ceph_assert(r == 0)
 ceph version Development (no_version) octopus (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x150) [0xffff843f9c04]
 2: (()+0x156ae00) [0xffff843f9e00]
 3: (ceph::mutex_debug_detail::mutex_debug_impl<false>::~mutex_debug_impl()+0x34) [0x6ae94c]
 4: (TestObjectRecorder::Handler::~Handler()+0x34) [0x75e1c0]
 5: (TestObjectRecorder::ObjectRecorderFlusher::~ObjectRecorderFlusher()+0xb4) [0x75adc4]
 6: (TestObjectRecorder_AppendFilledObject_Test::TestBody()+0x970) [0x755440]
 7: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x64) [0x79530c]
 8: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x50) [0x78e61c]
 9: (testing::Test::Run()+0xfc) [0x76d4f0]
 10: (testing::TestInfo::Run()+0x10c) [0x76ddf0]
 11: (testing::TestSuite::Run()+0x10c) [0x76e4cc]
 12: (testing::internal::UnitTestImpl::RunAllTests()+0x3dc) [0x778df0]
 13: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x64) [0x7965a4]
 14: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x50) [0x78f6c4]
 15: (testing::UnitTest::Run()+0xbc) [0x777848]
 16: (RUN_ALL_TESTS()+0x10) [0x69b6f0]
 17: (main()+0xf8) [0x69b2c4]
 18: (__libc_start_main()+0xe0) [0xffff82a9f8a0]
2019-08-04T02:22:12.660-0400 ffff82770000 -1 /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/common/mutex_debug.h: In function 'ceph::mutex_debug_detail::mutex_debug_impl<<anonymous> >::~mutex_debug_impl() [with bool Recursive = false]' thread ffff82770000 time 2019-08-04T02:22:12.630469-0400
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/common/mutex_debug.h: 111: FAILED ceph_assert(r == 0)
```

see https://jenkins.ceph.com/job/ceph-pull-requests-arm64/4168/consoleFull

File: src/test/journal/test_ObjectRecorder.cc [hunk 1]
```diff
@@ -70,6 +70,10 @@ class TestObjectRecorder : public RadosTestFixture {
 	C_SaferCond cond;
 	object_recorder->flush(&cond);
 	cond.wait();
+	std::scoped_lock l{*m_handler.object_lock};
+	if (!object_recorder) {
```
Comment (dillaman) [inline src/test/journal/test_ObjectRecorder.cc]: Should this really be testing that the pointer is null?

Comment (tchaikov) [reply src/test/journal/test_ObjectRecorder.cc]: no. and sorry, i pushed a wrong branch.

Comment (tchaikov) [reply src/test/journal/test_ObjectRecorder.cc]: fixed and repushed,

**Comment by tchaikov** [comment]
changelog

- should check if an object is already closed before closing it.

**Comment by tchaikov** [comment]
changelog

- should hold the object's lock instead of the Handler's lock.

**Comment by tchaikov** [comment]
`unittest_journal` is only exercised by `make check`. so i am merging this, since jenkins is happy with it.

**Pull request merged**